crates/nu-command/src/filesystem/cp.rs

```diff
@@ -352,17 +352,32 @@ fn copy_file(src: PathBuf, dst: PathBuf, span: Span) -> Value {
             Value::String { val: msg, span }
         }
         Err(e) => {
-            let message = format!("copy file {src:?} failed: {e}");
-
+            let message_src = format!("copy file {src:?} failed: {e}");
+            let message_dst = format!("copy file {dst:?} failed: {e}");
             use std::io::ErrorKind;
             let shell_error = match e.kind() {
-                ErrorKind::NotFound => ShellError::FileNotFoundCustom(message, span),
-                ErrorKind::PermissionDenied => ShellError::PermissionDeniedError(message, span),
-                ErrorKind::Interrupted => ShellError::IOInterrupted(message, span),
-                ErrorKind::OutOfMemory => ShellError::OutOfMemoryError(message, span),
+                ErrorKind::NotFound => {
+                    if std::path::Path::new(&dst).exists() {
+                        ShellError::FileNotFoundCustom(message_src, span)
+                    } else {
+                        ShellError::FileNotFoundCustom(message_dst, span)
+                    }
+                }
+                ErrorKind::PermissionDenied => match std::fs::metadata(&dst) {
+                    Ok(meta) => {
+                        if meta.permissions().readonly() {
+                            ShellError::PermissionDeniedError(message_dst, span)
+                        } else {
+                            ShellError::PermissionDeniedError(message_src, span)
+                        }
+                    }
+                    Err(_) => ShellError::PermissionDeniedError(message_dst, span),
+                },
```

Comment:
Thanks! 
(Not the absolute expert on the different platforms if there are edge cases but looks plausible)